hw10_decor_iter/hw_2_decor_iter.py

```diff
@@ -0,0 +1,270 @@
+
++3 дополнительных балла за реализацию того, чтобы унаследованные от str и set методы также возвращали объект интересующего нас класса (то есть MyString.replace(..., ...) должен возвращать MyString). Переопределять методы нельзя
+
+Модули использовать нельзя
+
+def change_outp_class():
+    def wrapp(func):
+        def inner_func(*args, **kwargs):
+            classname = type(args[0])
+            result = func(*args, **kwargs)
+            return classname(result)
+        return inner_func
+    return wrapp
+​
+​
+class MyString(str):
+    @change_outp_class()
+    def reverse(self):
+        return self[::-1]
+    
+    @change_outp_class()
+    def make_uppercase(self):
+        return "".join([chr(ord(char) - 32) if 97 <= ord(char) <= 122 else char for char in self])
+    
+    @change_outp_class()
+    def make_lowercase(self):
+        return "".join([chr(ord(char) + 32) if 65 <= ord(char) <= 90 else char for char in self])
+    
+    @change_outp_class()
+    def capitalize_words(self):
+        return " ".join([word.capitalize() for word in self.split()])
+    
+    
+class MySet(set):
+    def is_empty(self):
+        return len(self) == 0
+    
+    def has_duplicates(self):
+        return len(self) != len(set(self))
+    
+    @change_outp_class()
+    def union_with(self, other):
+        return self.union(other)
+    
+    def intersection_with(self, other):
+        return self.intersection(other)
+    
+    @change_outp_class()
+    def difference_with(self, other):
+        return self.difference(other)
+string_example = MyString("Aa Bb Cc")
+set_example_1 = MySet({1, 2, 3, 4})
+set_example_2 = MySet({3, 4, 5, 6, 6})
+​
+print(type(string_example.reverse()))
+print(type(string_example.make_uppercase()))
+print(type(string_example.make_lowercase()))
+print(type(string_example.capitalize_words()))
+print()
+print(type(set_example_1.is_empty()))
+print(type(set_example_2.has_duplicates()))
+print(type(set_example_1.union_with(set_example_2)))
+print(type(set_example_1.difference_with(set_example_2)))
+<class '__main__.MyString'>
+<class '__main__.MyString'>
+<class '__main__.MyString'>
+<class '__main__.MyString'>
+
+<class 'bool'>
+<class 'bool'>
+<class '__main__.MySet'>
+<class '__main__.MySet'>
+​
+Задание 4 (5 баллов)
+Напишите декоратор switch_privacy:
+
+Делает все публичные методы класса приватными
+Делает все приватные методы класса публичными
+Dunder методы и защищённые методы остаются без изменений
+Должен работать тестовый код ниже, в теле класса писать код нельзя
+Модули использовать нельзя
+
+def сhange_privacy(cls):
+    public_methods = [m for m in dir(cls) if not m.startswith('_')]
+    private_methods = [m for m in dir(cls) if m.startswith(f'_{cls.__name__}__')]
+        # print(public_methods,'\n', private_methods)
+​
+    for method in public_methods:
+        new_name = f'_{cls.__name__}__' + method
```

Comment:
f-строки уже намекают, что не надо складывать строки друг с другом :)
```suggestion
        new_name = f'_{cls.__name__}__{method}'
```